Sources/ComposableArchitectureMacros/ReducerMacro.swift

```diff
@@ -0,0 +1,172 @@
+      let identifier = binding.pattern.as(IdentifierPatternSyntax.self)?.identifier,
+      identifier.text == "body",
+      case .getter = binding.accessorBlock?.accessors
+    {
+      if let reduce = declaration.memberBlock.members.first(where: {
+        guard
+          let method = $0.decl.as(FunctionDeclSyntax.self),
+          method.name.text == "reduce",
+          method.signature.parameterClause.parameters.count == 2,
+          let state = method.signature.parameterClause.parameters.first,
+          state.firstName.text == "into",
+          state.type.as(AttributedTypeSyntax.self)?.specifier?.text == "inout",
+          method.signature.parameterClause.parameters.last?.firstName.text == "action",
+          method.signature.effectSpecifiers == nil,
+          method.signature.returnClause?.type.as(IdentifierTypeSyntax.self) != nil
+        else {
+          return false
+        }
+        return true
+      }) {
+        context.diagnose(
+          Diagnostic(
+            node: reduce.decl.cast(FunctionDeclSyntax.self).name,
+            message: MacroExpansionWarningMessage(
+              """
+              A 'reduce' method should not be defined in a reducer with a 'body'; it takes \
+              precedence and 'body' will never be invoked
+              """
+            ),
+            notes: [
+              Note(
+                node: Syntax(identifier),
+                message: MacroExpansionNoteMessage("'body' defined here")
+              )
+            ]
+          )
+        )
+      }
+      for attribute in property.attributes {
+        guard
+          case let .attribute(attribute) = attribute,
+          let attributeName = attribute.attributeName.as(IdentifierTypeSyntax.self)?.name.text
+        else { continue }
+        guard
+          !attributeName.starts(with: "ReducerBuilder"),
+          !attributeName.starts(with: "ComposableArchitecture.ReducerBuilder")
+        else { return [] }
+      }
+      return [
+        AttributeSyntax(
+          attributeName: IdentifierTypeSyntax(
+            name: .identifier("ComposableArchitecture.ReducerBuilder<Self.State, Self.Action>")
+          )
+        )
+      ]
+    } else {
+      return []
+    }
+  }
+}
+
+extension DeclGroupSyntax {
+  var inheritanceClause: InheritanceClauseSyntax? {
+    if let decl = self.as(StructDeclSyntax.self) {
+      return decl.inheritanceClause
+    } else if let decl = self.as(ClassDeclSyntax.self) {
+      return decl.inheritanceClause
+    } else if let decl = self.as(EnumDeclSyntax.self) {
+      return decl.inheritanceClause
+    } else {
+      return nil
+    }
+  }
+
+  var memberBlock: MemberBlockSyntax? {
+    if let decl = self.as(StructDeclSyntax.self) {
+      return decl.memberBlock
+    } else if let decl = self.as(ClassDeclSyntax.self) {
+      return decl.memberBlock
+    } else if let decl = self.as(EnumDeclSyntax.self) {
+      return decl.memberBlock
+    } else {
+      return nil
+    }
+  }
+}
```

Comment:
I'm not sure if you've already seen and intentionally ignored it due to certain reasons, or if you've missed it, but in the `swift-syntax` world, we have `SyntaxTraits`. You could use the `DeclGroupSyntax` here:

https://github.com/apple/swift-syntax/blob/39e732e277b47a19b4f057cec971892d4c757156/Sources/SwiftSyntax/generated/SyntaxTraits.swift#L66-L92

It wouldn't provide the exact same logic as this code, because types other than `Struct`, `Enum`, and `Class` can also be `DeclGroupSyntax`. However, I just wanted to let you know about the existence of such traits in case you weren't aware 🫶

Reply:
Huh, I don't remember exactly. I either missed this or it wasn't working in the way I was first trying it. But looks like I can get rid of these helpers now, thanks! While the macro can't apply to those other contexts, I suppose the compiler will fail in a different way.